### PR TITLE
[SYCL-MLIR] Lower all SYCL operations to LLVM

### PR DESCRIPTION
**Pull request by victor-eds**
Before this, we could not lower some SYCL operations to the LLVM dialect, i.e., SYCL grid operations, `sycl.group.get_local_id` and `sycl.group.get_local_linear_id`. Now, using conversion to GPU and SPIRV patterns, we can.

Also, the arith operations introduced by the conversion are also lowered to LLVM.

### CODE REVIEW

File: mlir-sycl/include/mlir/Conversion/SYCLPasses.td [hunk 1]
```diff
@@ -21,7 +21,14 @@ def ConvertSYCLToLLVM : Pass<"convert-sycl-to-llvm", "ModuleOp"> {
...
     TODO: add docs referenced above.
   }];
-  let dependentDialects = ["LLVM::LLVMDialect"];
+  let dependentDialects = [
+      "LLVM::LLVMDialect",
+      "arith::ArithDialect",
+      "gpu::GPUDialect",
+      "memref::MemRefDialect",
+      "spirv::SPIRVDialect",
+      "vector::VectorDialect"
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Conversion/SYCLPasses.td:30]: We will be generating operations from way more dialects as intermediate steps.

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-to-llvm.mlir [hunk 2]
```diff
@@ -1061,31 +1061,89 @@ func.func @test(%group: memref<?x!sycl_group_1_>, %i: i32) -> i64 {
...
+  gpu.module @kernels {
 // CHECK-LABEL:   llvm.func @test(
 // CHECK-SAME:                    %[[VAL_0:.*]]: !llvm.ptr<[[GROUP3:.*]]>) -> !llvm.[[ID3:.*]] {
-// CHECK-NEXT:       %[[VAL_1:.*]] = arith.constant 1 : i32
+// CHECK-NEXT:       %[[VAL_1:.*]] = llvm.mlir.constant(1 : i32) : i32
 // CHECK-NEXT:       %[[VAL_2:.*]] = llvm.alloca %[[VAL_1]] x !llvm.[[ID3]] : (i32) -> !llvm.ptr<[[ID3]]>
 // CHECK-NEXT:       %[[VAL_3:.*]] = llvm.getelementptr inbounds %[[VAL_2]][0, 0, 0, 0] : (!llvm.ptr<[[ID3]]>) -> !llvm.ptr<i64>
-// CHECK-NEXT:       %[[VAL_4:.*]] = arith.constant 0 : i32
-// CHECK-NEXT:       %[[VAL_5:.*]] = sycl.local_id %[[VAL_4]] : index
-// CHECK-NEXT:       %[[VAL_6:.*]] = arith.index_castui %[[VAL_5]] : index to i64
```
Comment (victor-eds) [inline mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-to-llvm.mlir:1071]: We are lowering these three operations to what we see below.

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-to-llvm.mlir [hunk 3]
```diff
@@ -1061,31 +1061,89 @@ func.func @test(%group: memref<?x!sycl_group_1_>, %i: i32) -> i64 {
 !sycl_range_3_ = !sycl.range<[3], (!sycl.array<[3], (memref<3xi64, 4>)>)>
 !sycl_group_3_ = !sycl.group<[3], (!sycl_range_3_, !sycl_range_3_, !sycl_range_3_, !sycl_id_3_)>
 
+module attributes {gpu.container} {
+  gpu.module @kernels {
```
Comment (victor-eds) [inline mlir-sycl/test/Conversion/SYCLToLLVM/sycl-methods-to-llvm.mlir:1065]: We need to wrap this as we see to be able to lower to `spirv` in the intermediate step.

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 4]
```diff
@@ -2050,22 +2063,128 @@ void mlir::populateSYCLToLLVMConversionPatterns(
...
+  LLVM_DEBUG(llvm::dbgs() << "Lowering to SPIRV...\n");
+
+  auto &context = getContext();
+  auto module = getOperation();
+
+  const auto res = failure(
+      module
+          .walk([&context](gpu::GPUModuleOp gpuModule) {
+            // We walk the different GPU modules looking for different SPIRV
+            // target
```
Comment (etiotto) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2099]: nit: comment can span less lines

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 5]
```diff
@@ -2050,22 +2063,128 @@ void mlir::populateSYCLToLLVMConversionPatterns(
...
+                  SYCLWorkGroupIDOp, SYCLNumWorkItemsOp, SYCLWorkGroupSizeOp,
+                  SYCLLocalIDOp, SYCLGlobalIDOp, SYCLSubGroupIDOp,
+                  SYCLNumSubGroupsOp, SYCLSubGroupSizeOp,
+                  // Not convertible to GPU dialect
+                  SYCLGlobalOffsetOp, SYCLNumWorkGroupsOp,
+                  SYCLSubGroupLocalIDOp, SYCLSubGroupMaxSizeOp>(op);
+            });
+
+            // Add generic source and target materializations to handle cases
+            // where
```
Comment (etiotto) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2136]: nit: comment can span less lines

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 6]
```diff
@@ -2077,14 +2196,55 @@ void ConvertSYCLToLLVMPass::runOnOperation() {
...
+          << "Module after reconciling UnrealizedConversionCasts operations:\n";
+      module.dump(););
+}
 
-  target.addLegalOp<ModuleOp>();
-  if (failed(applyPartialConversion(module, target, std::move(patterns))))
-    signalPassFailure();
+void ConvertSYCLToLLVMPass::runOnOperation() {
+  constexpr int requiredRuns{2};
+  for (int i = 0; i < requiredRuns; ++i) {
```
Comment (etiotto) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2241]: How do we know we need exactly 2 passes ? Also shouldn't we convert from SYCL to SPIRV first and then convert from to LLVM after ?

**Comment by victor-eds** [comment]
**NOTE**: We're just missing tests for grid ops. Functionally complete.

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 4 above)
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp]: Fixed

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 6 above)
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2239]: Comment in file:
```
 2065 /// A pass converting MLIR SYCL operations into LLVM dialect.
 2066 ///
 2067 /// This pass relies on SYCL to GPU and target dialects, e.g., SPIRV, conversion
 2068 /// patterns. This pass is executed in 5 steps:
 2069 /// 1. Convert grid ops to a target dialect, e.g., SPIRV;
 2070 /// 2. Lower all of the operations to LLVM. As some of the operations will yield
 2071 ///    SYCL grid ops, we need to perform the same conversions as in step 1
 2072 ///    again;
 2073 /// 3. Same as 1;
 2074 /// 4. Lower remaining operations to LLVM;
 2075 /// 5. Resolve UnrealizedConversionCastOps appearing due to the fact that this
 2076 ///    pass is performed in several steps.
```

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 7]
```diff
@@ -2077,14 +2194,55 @@ void ConvertSYCLToLLVMPass::runOnOperation() {
...
+    if (convertToSPIRV().failed() || convertToLLVM().failed()) {
```
Comment (sommerlukas) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2240]: IIUC, the second conversion to LLVM would still allow `SYCLLocalIDOp` as a legal operation, but we want to completely lower to LLVM dialect in that step. Should we enforce this by making the operation illegal on the second run?

Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp]: That's a good point. I've changed that. Thanks!

**Review by etiotto — APPROVED** [review]

**Review by sommerlukas — APPROVED** [review]

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 8]
```diff
@@ -2050,22 +2063,128 @@ void mlir::populateSYCLToLLVMConversionPatterns(
 
 namespace {
 /// A pass converting MLIR SYCL operations into LLVM dialect.
+///
+/// This pass relies on SYCL to GPU and target dialects, e.g., SPIRV, conversion
+/// patterns. This pass is executed in 5 steps:
+/// 1. Convert grid ops to a target dialect, e.g., SPIRV;
+/// 2. Lower all of the operations to LLVM. As some of the operations will yield
+///    SYCL grid ops, we need to perform the same conversions as in step 1
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2071]: What kind of operations will yield SYCL grid ops? 
Is is possible to lower to LLVM first, then to SPIRV, then LLVM again? (saving one invocation to SPIRV)

File: mlir-sycl/lib/Conversion/SYCLToSPIRV/SYCLToSPIRV.cpp [hunk 9]
```diff
@@ -199,9 +199,11 @@ class GridOpPatternNoIndex : public GridOpPattern<OpTy> {
 void rewrite1D(Operation *op, spirv::BuiltIn builtin,
                const SPIRVTypeConverter &typeConverter,
                ConversionPatternRewriter &rewriter) {
-  rewriter.replaceOp(op,
-                     spirv::getBuiltinVariableValue(
-                         op, builtin, typeConverter.getIndexType(), rewriter));
+  const auto res = spirv::getBuiltinVariableValue(
+      op, builtin, typeConverter.getIndexType(), rewriter);
+  rewriter.replaceOp(op, convertScalarToDtype(rewriter, op->getLoc(), res,
+                                              op->getResultTypes()[0],
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Conversion/SYCLToSPIRV/SYCLToSPIRV.cpp:205]: Should we convert to the converted type using `typeConverter()->convertType()`?

Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToSPIRV/SYCLToSPIRV.cpp]: Well, that should be the very same type everytime, as the result is `i32`, but I guess it doesn't hurt to be extra cautious. Added.

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 10]
```diff
@@ -2203,9 +2203,16 @@ LogicalResult ConvertSYCLToLLVMPass::convertToLLVM(int currentIter) {
   populateSPIRVToLLVMFunctionConversionPatterns(converter, patterns);
 
   LLVMConversionTarget target(context);
-  if (currentIter != requiredRuns - 1) {
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp]: `constexpr static int requiredRuns{2};` can be removed too.

**Review by whitneywhtsang — APPROVED** [review]

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 8 above)
Comment (whitneywhtsang) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2071]: opps, the comment needs to be updated.

Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp]: I'm super clumsy today :sweat_smile:

**Pull request merged**